python/pyarrow/serialization.py

```diff
@@ -42,15 +43,24 @@ def _serialize_numpy_array_list(obj):
         # the view.
         if not obj.flags.c_contiguous:
             obj = np.ascontiguousarray(obj)
-        return obj.view('uint8'), obj.dtype.str
+
+        return obj.view('uint8'), pickle.dumps(obj.dtype)
     else:
         return obj.tolist(), obj.dtype.str
 
 
 def _deserialize_numpy_array_list(data):
     if data[1] != '|O':
         assert data[0].dtype == np.uint8
-        return data[0].view(data[1])
+
+        try:
+            dtype = pickle.loads(data[1])
+        except (TypeError, pickle.UnpicklingError):
```

Comment:
I don't think we should try to unpickle data that we know isn't a pickle (in extreme cases there may even be false positives, though that's unlikely). Rather, perhaps we should change the serialization format for Numpy arrays. For example a 3-tuple `(kind, serialized dtype, data)`. The `kind` (for example a string: "object", "struct", "scalar") would help discriminate between different serialization formats. After all we already have a distinct deserialization path for object arrays.